paasta_tools/cli/cmds/spark_run.py

```diff
@@ -522,6 +511,47 @@ def should_enable_compact_bin_packing(disable_compact_bin_packing, cluster_manag
     return True
 
 
+# inplace add a low priority podAffinityTerm for compact bin packing
+def add_compact_bin_packing_affinity_term(pod: Dict, spark_pod_label: str):
+    pod["metadata"]["labels"]["spark"] = spark_pod_label
+    pod["spec"]["affinity"]["podAffinity"].setdefault(
+        "preferredDuringSchedulingIgnoredDuringExecution", []
+    ).append(
+        {
+            "weight": 95,
+            "podAffinityTerm": {
+                "labelSelector": {
+                    "matchExpressions": [
+                        {
+                            "key": "spark",
+                            "operator": "In",
+                            "values": [spark_pod_label],
+                        }
+                    ]
+                },
+                "topologyKey": "topology.kubernetes.io/hostname",
+            },
+        }
+    )
+    return pod
+
+
+# inplace add nodeAffinity for node selection
+def add_node_affinity_terms(pod: Dict, instance_types: str):
+    pod["spec"]["affinity"]["nodeAffinity"][
+        "requiredDuringSchedulingIgnoredDuringExecution"
+    ].setdefault("nodeSelectorTerms", []).extend(
+        [
+            {
+                "key": "node.kubernetes.io/instance-type",
```

Comment:
just curious: do we want users to specify an instance type (e.g., g4dn.xlarge vs g4dn.2xlarge) or would we be fine having them specify a family (e.g., g4dn) and letting karpenter spin up the most optimal instance type for the given requests at the time?